apis/v1alpha3/clusterworkloadresourcemapping_types.go

```diff
@@ -18,24 +18,49 @@ package v1alpha3
 
 import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
-// ClusterWorkloadResourceMappingVersion defines the mapping for a specific version of an workload resource.
-type ClusterWorkloadResourceMappingVersion struct {
+// ClusterWorkloadResourceMappingTemplate defines the mapping for a specific version of an workload resource to a
+// logical PodTemplateSpec-like structure.
+type ClusterWorkloadResourceMappingTemplate struct {
 	// Version is the version of the workload resource that this mapping is for.
 	Version string `json:"version"`
-	// Containers is the collection of JSONPaths that container configuration may be written to.
-	Containers []string `json:"containers,omitempty"`
-	// Envs is the collection of JSONPaths that env configuration may be written to.
-	Envs []string `json:"envs,omitempty"`
-	// VolumeMounts is the collection of JSONPaths that volume mount configuration may be written to.
-	VolumeMounts []string `json:"volumeMounts,omitempty"`
-	// Volumes is the JSONPath that volume configuration must be written to.
+	// Annotations is a Restricted JSONPath that references the annotations map within the workload resource. These
+	// annotations must end up in the resulting Pod, and are generally not the workload resource's annotations.
+	// Defaults to `.spec.template.metadata.annotations`.
+	Annotations string `json:"annotations"`
+	// Containers is the collection of mappings to container-like fragments of the workload resource. Defaults to
+	// mappings appropriate for a PodSpecable resource.
+	Containers []ClusterWorkloadResourceMappingContainer `json:"containers,omitempty"`
+	// Volumes is a Restricted JSONPath that references the slice of volumes within the workload resource. Defaults to
+	// `.spec.template.spec.volumes`.
 	Volumes string `json:"volumes"`
 }
 
+// ClusterWorkloadResourceMappingContainer defines the mapping for a specific fragment of an workload resource
+// to a Container-like structure.
+//
+// Each mapping defines exactly one path that may match multiple container-like fragments within the workload
+// resource. For each object matching the path the name, env and volumeMounts expressions are resolved to find those
+// structures.
+type ClusterWorkloadResourceMappingContainer struct {
+	// Path is the JSONPath within the workload resource that matches an existing fragment that is container-like.
+	Path string `json:"path"`
```

Comment:
it should be declared as non-empty

Reply:
hm, I see that is it required, but I thought that we want to add `+kubebuilder:validation:MinLength=1` on this field.